Uncaught Exception

Whilst running Psalm with the PHPUnit plugin on my project, at some point I caused it to break hard although I'm not sure from what exactly:

Full stack trace:

```
PHP Warning:  assert(): assert(null !== $param->type) failed in /path/to/project/vendor/psalm/plugin-phpunit/src/Hooks/TestCaseHandler.php on line 397
Warning: assert(): assert(null !== $param->type) failed in /path/to/project/vendor/psalm/plugin-phpunit/src/Hooks/TestCaseHandler.php on line 397
░Uncaught Exception: Argument 2 passed to Psalm\PhpUnitPlugin\Hooks\TestCaseHandler::Psalm\PhpUnitPlugin\Hooks\{closure}() must be an instance of Psalm\Type\Union, null given, called in /path/to/project/vendor/psalm/plugin-phpunit/src/Hooks/TestCaseHandler.php on line 415
Stack trace in the forked worker:
#0 /path/to/project/vendor/psalm/plugin-phpunit/src/Hooks/TestCaseHandler.php(415): Psalm\PhpUnitPlugin\Hooks\TestCaseHandler::Psalm\PhpUnitPlugin\Hooks\{closure}(Object(Psalm\Type\Union), NULL, false, 0)
#1 /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ClassAnalyzer.php(963): Psalm\PhpUnitPlugin\Hooks\TestCaseHandler::afterStatementAnalysis(Object(PhpParser\Node\Stmt\Class_), Object(Psalm\Storage\ClassLikeStorage), Object(Psalm\Internal\Analyzer\ClassAnalyzer), Object(Psalm\Codebase), Array)
#2 /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/FileAnalyzer.php(211): Psalm\Internal\Analyzer\ClassAnalyzer->analyze(Object(Psalm\Context), Object(Psalm\Context))
#3 /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(340): Psalm\Internal\Analyzer\FileAnalyzer->analyze(NULL)
#4 /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Fork/Pool.php(193): Psalm\Internal\Codebase\Analyzer->Psalm\Internal\Codebase\{closure}(4, '/Users/tfidry/P...')
#5 /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(406): Psalm\Internal\Fork\Pool->__construct(Array, Object(Closure), Object(Closure), Object(Closure), Object(Closure))
#6 /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(269): Psalm\Internal\Codebase\Analyzer->doAnalysis(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 3)
#7 /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ProjectAnalyzer.php(639): Psalm\Internal\Codebase\Analyzer->analyzeFiles(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 3, false, true)
#8 /path/to/project/vendor/vimeo/psalm/src/psalm.php(676): Psalm\Internal\Analyzer\ProjectAnalyzer->check('/Users/tfidry/P...', true)
#9 /path/to/project/vendor/vimeo/psalm/psalm(2): require_once('/Users/tfidry/P...')
#10 {main} in /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Fork/Pool.php:357
Stack trace:
#0 /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Fork/Pool.php(389): Psalm\Internal\Fork\Pool->readResultsFromChildren()
#1 /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(473): Psalm\Internal\Fork\Pool->wait()
#2 /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Codebase/Analyzer.php(269): Psalm\Internal\Codebase\Analyzer->doAnalysis(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 3)
#3 /path/to/project/vendor/vimeo/psalm/src/Psalm/Internal/Analyzer/ProjectAnalyzer.php(639): Psalm\Internal\Codebase\Analyzer->analyzeFiles(Object(Psalm\Internal\Analyzer\ProjectAnalyzer), 3, false, true)
#4 /path/to/project/vendor/vimeo/psalm/src/psalm.php(676): Psalm\Internal\Analyzer\ProjectAnalyzer->check('/Users/tfidry/P...', true)
#5 /path/to/project/vendor/vimeo/psalm/psalm(2): require_once('/Users/tfidry/P...')
#6 {main}
(Psalm 4.3.1@2feba22a005a18bf31d4c7b9bdb9252c73897476 crashed due to an uncaught Throwable)
make: *** [psalm] Error 1
```

PHPUnit 9.4.4 & Psalm 4.3.1

--- a/src/Hooks/TestCaseHandler.php
+++ b/src/Hooks/TestCaseHandler.php
@@ -393,9 +393,9 @@ class TestCaseHandler implements
                         }
                         $potential_argument_type = $potential_argument_types[$param_offset];
 
-                        assert(null !== $param->type);
+                        $param_type = $param->type === null ? Type::getMixed() : $param->type;
                         if ($param->is_variadic) {
-                            $param_types = self::getAtomics($param->type);
+                            $param_types = self::getAtomics($param_type);
                             $variadic_param_type = new Type\Union(array_values($param_types));
 
                             // check remaining argument types
@@ -411,7 +411,7 @@ class TestCaseHandler implements
                             break;
                         }
 
-                        $checkParam($potential_argument_type, $param->type, $param->is_optional, $param_offset);
+                        $checkParam($potential_argument_type, $param_type, $param->is_optional, $param_offset);
                     }
                 }
             }

--- a/tests/acceptance/TestCase.feature
+++ b/tests/acceptance/TestCase.feature
@@ -1048,6 +1048,26 @@ Feature: TestCase
       | Type            | Message                    |
       | InvalidDocblock | %@psalm-force-push-master% |
 
+  Scenario: Missing param type on a test with tuple data provider does not crash psalm
+    Given I have the following code
+      """
+      class MyTestCase extends TestCase {
+        /**
+         * @dataProvider provide
+         * @psalm-suppress MissingParamType
+         */
+        public function testSomething($data): void {}
+
+        /** @return iterable<string, array{int}> */
+        public function provide(): iterable {
+          return ['case 1' => [1]];
+        }
+      }
+      """
+    When I run Psalm
+    Then I see exit code 0
+    And I see no errors
+
   Scenario: Providers referenced in shorthand docblocks are not marked as unused
     Given I have the following code
       """

